### PR TITLE
Composition errors tracked through services AND adding service validations

### PR DESCRIPTION
**Pull request by KensoDev**
In order to track errors as we go ahead and traverse the graph of
services we need to transform, we need to track errors in the
composition.

Right before you echo the composition to YAML via stdout, you should
echo to stderr if there are any errors.

For example: if one of your plugins requires a validation step. this
will error out until you fix it (eg: botoenv, placement of a file and
more)

### CODE REVIEW

File: docker_etude/models/composition.py [hunk 1]
```diff
@@ -36,16 +36,26 @@ def __init__(self,
...
+            self.add_error(error_message)
+
     def add_volume(self, volume):
         self.volumes[volume.name] = volume
 
+    def add_error(self, service, error_message):
+        self.errors.append("{}: {}".format(service.name, error_message))
+
+    def has_errors(self):
+        return len(self.errors) > 0
```
Comment (jessemyers) [inline docker_etude/models/composition.py]: Collection size is truthy; just `return self.errors` or possibly `return bool(self.errors)`

File: docker_etude/models/service.py [hunk 2]
```diff
@@ -49,6 +49,13 @@ def __init__(self,
         self.ports = ports or []
         self.ulimits = ulimits or {}
         self.volumes = volumes or []
+        self.errors = []
+
+    def add_error(self, error_message):
+        self.errors.append(error_message)
+
+    def has_errors(self):
```
Comment (jessemyers) [inline docker_etude/models/service.py]: Collection size is truthy; just `return self.errors` or possibly `return bool(self.errors)` or even omit entirely and let the callers check `service.errors`

File: docker_etude/models/composition.py [hunk 3]
```diff
@@ -36,16 +36,26 @@ def __init__(self,
...
         self.volumes = volumes or OrderedDict()
+        self.errors = []
 
     def add_network(self, network):
         self.networks[network.name] = network
 
     def add_service(self, service):
         self.services[service.name] = service
 
+        for error_message in service.errors:
```
Comment (jessemyers) [inline docker_etude/models/composition.py]: I think you'd be better off without saving errors in the composition if the use case is to track them for services.

Instead, I'd focus on writing a generator or `@property` on the composition that exposes underlying errors. 

Either:

```
def iter_errors(self):
    for service in self.services:
        yield from self.service.errors
```

Or:

```
@property
def errors(self):
    return [
        error
        for service in self.services
        for error in service.errors
   ]
```

File: docker_etude/models/composition.py [hunk 4]
```diff
@@ -36,16 +36,26 @@ def __init__(self,
...
+            self.add_error(error_message)
+
     def add_volume(self, volume):
         self.volumes[volume.name] = volume
 
+    def add_error(self, service, error_message):
+        self.errors.append("{}: {}".format(service.name, error_message))
```
Comment (jessemyers) [inline docker_etude/models/composition.py]: I don't think you want message formatting here; I think that's the job of the CLI/UI layer. If you want to know where errors are coming from, I'd either create an `Error` object or use a tuple of `(thing, message)`

File: docker_etude/models/validation.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+class FileExistValidation(Validation):
+    def __init__(self, filepath):
+        self.filepath = filepath
+
+    def is_valid(self):
+        valid_file = Path(self.filepath)
+        return valid_file.is_file()
+
+
+class ServiceValiation(Model):
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: spelling: Validation

File: docker_etude/models/validation.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+These validations will support multiple items, currently implemented is "file"
+
+"""
+
+from abc import ABC, abstractclassmethod, abstractmethod
+from docker_etude.models.base import Model
+from pathlib import Path
+
+class Validation(ABC):
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: `Validator` or `Validation`

File: docker_etude/models/validation.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    @abstractmethod
+    def is_valid():
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: If you go with `Validator`, this could just be `__call__`

File: docker_etude/models/validation.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                yield validation
+
+    @abstractmethod
+    def name(cls):
+        pass
+
+    @classmethod
+    def get(self, name):
+        """
+        Getsthe validation for the name
```
Comment (tomgoren) [inline docker_etude/models/validation.py]: `Gets the`

File: docker_etude/models/composition.py [hunk 9]
```diff
@@ -43,9 +43,23 @@ def add_network(self, network):
     def add_service(self, service):
         self.services[service.name] = service
 
+        for error_message in service.errors:
```
Comment (jessemyers) [inline docker_etude/models/composition.py]: Do we need L46-L48 if we also have the `iter_service_errors` function?

File: docker_etude/models/composition.py [hunk 10]
```diff
@@ -43,9 +43,23 @@ def add_network(self, network):
...
         self.volumes[volume.name] = volume
 
+    def iter_service_errors(self):
+        for service in self.services:
+            for error in service.errors:
+                yield error
+
+    @property
+    def errors(self):
+        return list(
```
Comment (jessemyers) [inline docker_etude/models/composition.py]: I think you can just `return list(self.iter_service_errors())`

File: docker_etude/models/service.py [hunk 11]
```diff
@@ -7,6 +7,21 @@
...
 
 
+class ServiceError(Exception):
+    def __init__(self, service, error_message):
+        self.service = service
+        self.error_message = error_message
+
+    @property
+    def normalized_message(self):
+        print(self.service.to_safe_dict())
```
Comment (jessemyers) [inline docker_etude/models/service.py]: Do we want these prints?

Comment (KensoDev) [inline docker_etude/models/service.py]: no

File: docker_etude/models/validation.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+        :name: name of a validation
+        :returns: single validation that match the name or EmptyValidation
+
+        """
+        matching = list(BaseValidation.iter_matching_validations(name))
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: More idiomatic:

```
try:
    return next(BaseValidation.iter_matching_validations(name))
except StopIteration:
    return EmptyValidation
```

File: docker_etude/models/validation.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    """
+    Check if a File exists in a specific path
+
+    """
+    @classmethod
+    def name(cls):
+        return "file"
+
+    def __init__(self, **kwargs):
+        print(kwargs)
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: print?

File: docker_etude/models/composition.py (hunk 10 above)
Comment (KensoDev) [inline docker_etude/models/composition.py]: Yes, you are right. I missed a cleanup here. 
I wanted to return the message but decided on returning the object so the cli can decide on formatting

File: docker_etude/models/validation.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        for service_validation in self.iter_matching_validations(service):
+            for validation in service_validation.validations:
+                if not validation.validate():
+                    passing = False
+                    service.add_error("Failed to pass validation: {}".format(type(validation).name()))
+
+        return passing
+
+    def iter_matching_validations(self, service):
+        service_validations = list(filter(lambda x: x.service == service.name, self.validations))
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: a comprehension is preferable to a filter here.

File: docker_etude/models/validation.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+        for key in dct.keys():
+            sv = ServiceValidations(
+                service=key,
+                validations=list(
+                    BaseValidation.from_dict(v)
+                    for v in dct[key]
+                )
+            )
+
+            validations.append(sv)
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: Use a comprehension instead of `list.append()`

File: docker_etude/models/validation.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+    @classmethod
+    def from_dict(cls, dct):
+        validations = []
+
+        for key in dct.keys():
+            sv = ServiceValidations(
+                service=key,
+                validations=list(
+                    BaseValidation.from_dict(v)
+                    for v in dct[key]
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: spell out variables: `for value in dct[key]`

File: docker_etude/models/composition.py (hunk 9 above)
Comment (KensoDev) [inline docker_etude/models/composition.py]: No, we don't
Will clean up this part.

**Comment by KensoDev** [comment]
I am not 100% happy with the internal `validation` classes. External API is good IMHO.

File: docker_etude/models/validation.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+from os import getenv
+from pathlib import Path
+
+
+class BaseValidation(ABC):
+    def __init__(self, **kwargs):
+        pass
+
+    @classmethod
+    def all_validations(self):
```
Comment (jessemyers) [inline docker_etude/models/validation.py]: I think you can also do:

```
return BaseValidation.__subclasses__()
```

That's nice an extensible.

Comment (KensoDev) [inline docker_etude/models/validation.py]: ohhhh. tricks

**Pull request merged**